File: patterns/cancel-activity/rules-js/cancel-activity.js
let dimTimer = null;
let continueDimming = true;

rules.JSRule({
  name: "Start Dimming",
  triggers: [triggers.ItemCommandTrigger("StartDimTrigger", "ON")],
  execute: (event) => {
    continueDimming = true;

    function step() {
      const currLevel = isNaN(parseInt(items.getItem("DimLamp").state)) ? 0 : parseInt(items.getItem("DimLamp").state);
      const target = isNaN(parseInt(items.getItem("DimTarget").state)) ? 0 : parseInt(items.getItem("DimTarget").state);

      if (currLevel >= target || !continueDimming) {
        console.log("Dimmen beendet");
        dimTimer = null;
        return;
      }

      items.getItem("DimLamp").sendCommand(currLevel + 1);
      dimTimer.reschedule(time.ZonedDateTime.now().plusSeconds(1));
    }

    dimTimer = actions.ScriptExecution.createTimer(time.ZonedDateTime.now(), step);
  }
});

rules.JSRule({
  name: "Cancel Dimming",
  triggers: [triggers.ItemCommandTrigger("CancelDimTrigger", "ON")],
  execute: (event) => {
    continueDimming = false;
  }
});

File: patterns/cancel-activity/rules-python/cancel_activity.py
import threading

from openhab import rule, Registry, logger
from openhab.triggers import ItemCommandTrigger

dim_timer = None
continue_dimming = True


def step():
    global dim_timer, continue_dimming

    lamp_state = str(Registry.getItem("DimLamp").getState())
    target_state = str(Registry.getItem("DimTarget").getState())
    curr_level = int(lamp_state) if lamp_state not in ("NULL", "UNDEF") else 0
    target = int(target_state) if target_state not in ("NULL", "UNDEF") else 0

    if curr_level >= target or not continue_dimming:
        logger.info("Dimmen beendet bei {}%".format(curr_level))
        dim_timer = None
        return

    Registry.getItem("DimLamp").sendCommand(curr_level + 1)
    dim_timer = threading.Timer(1, step)
    dim_timer.start()


@rule(triggers=[ItemCommandTrigger("StartDimTrigger", "ON")])
class StartDimming:
    def execute(self, module, input):
        global continue_dimming
        continue_dimming = True
        self.logger.info("Dimmen gestartet")
        step()


@rule(triggers=[ItemCommandTrigger("CancelDimTrigger", "ON")])
class CancelDimming:
    def execute(self, module, input):
        global continue_dimming
        continue_dimming = False
        self.logger.info("Dimmen wird abgebrochen")
